File: src/main/java/practice/account/domain/TransactionHistoryList.java
package practice.account.domain;

import static practice.account.domain.TransactionType.PAYMENT;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

public record TransactionHistoryList(List<TransactionHistory> transactionHistoryList) {

  private static final BigDecimal DAILY_WITHDRAWAL_LIMIT = BigDecimal.valueOf(3_000_000);

  public boolean canWithdraw(BigDecimal money) {
    LocalDate today = LocalDate.now();
    BigDecimal totalWithdrawal = transactionHistoryList.stream()
        .filter(transactionHistory ->TransactionType.WITHDRAW.equals(transactionHistory.getType()))
        .filter(transactionHistory -> today.equals(transactionHistory.getTransactionDate()))
        .map(TransactionHistory::getAmount)
        .reduce(BigDecimal.ZERO, BigDecimal::add);
    return totalWithdrawal.add(money).compareTo(DAILY_WITHDRAWAL_LIMIT) > 0;
  }

  public void add(TransactionType type, BigDecimal amount, AccountInfo accountInfo) {
    transactionHistoryList.add(TransactionHistory.builder()
        .transactionDateTime(LocalDateTime.now())
        .type(type)
        .amount(amount)
        .counterPartyBank(accountInfo.getBank())
        .counterPartyAccountId(accountInfo.getAccountId())
        .build());
  }

  public void add(TransactionType type, BigDecimal amount, Long id) {
    transactionHistoryList.add(TransactionHistory.builder()
        .transactionDateTime(LocalDateTime.now())
        .type(type)
        .amount(amount)
        .counterPartyAccountId(PAYMENT.equals(type) ? null : String.valueOf(id))
        .paymentId(PAYMENT.equals(type) ? id : null)
        .build());
  }

  public void add(TransactionHistory transactionHistory) {
    transactionHistoryList.add(transactionHistory);
  }
}

File: src/main/kotlin/practice/account/domain/TransactionHistoryList.kt
package practice.account.domain

import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime

data class TransactionHistoryList(
    var items: MutableList<TransactionHistory>
) {

    constructor() : this(mutableListOf())

    fun canWithdraw(money: BigDecimal): Boolean {
        val today = LocalDate.now()
        val totalWithdrawal = items.stream()
            .filter { transactionHistory: TransactionHistory -> TransactionType.WITHDRAW == transactionHistory.type }
            .filter { transactionHistory: TransactionHistory -> today == transactionHistory.transactionDate() }
            .map { obj: TransactionHistory -> obj.amount }
            .reduce(BigDecimal.ZERO) { obj: BigDecimal?, augend: BigDecimal? -> obj?.add(augend) }
        return totalWithdrawal?.add(money)?.compareTo(DAILY_WITHDRAWAL_LIMIT)!! > 0
    }

    fun add(type: TransactionType, amount: BigDecimal, accountInfo: AccountInfo) {
        items.add(
            TransactionHistory(
                transactionDateTime = LocalDateTime.now(),
                type = type,
                amount = amount,
                counterPartyBank = accountInfo.bank,
                counterPartyAccountId = accountInfo.accountId
            )
        )
    }

    fun add(type: TransactionType, amount: BigDecimal, id: Long) {
        items.add(
            TransactionHistory(
                transactionDateTime = LocalDateTime.now(),
                type = type,
                amount = amount,
                counterPartyAccountId = if (TransactionType.PAYMENT == type) null else id.toString(),
                paymentId = if (TransactionType.PAYMENT == type) id else null
            )
        )
    }

    fun add(transactionHistory: TransactionHistory) {
        items.add(transactionHistory)
    }

    companion object {
        private val DAILY_WITHDRAWAL_LIMIT: BigDecimal = BigDecimal.valueOf(3000000)
    }
}
